<a href="https://colab.research.google.com/github/HChandeepa/Lung_Sound_Classification_System/blob/Master/Audio_Augmentation_(Pitch_Shifting).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Counts of Audio files Before augmentation

In [ ]:
import os

In [ ]:
# List of folder paths
folders = [
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Asthma',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/COPD',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia',
    '/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI'
]

In [ ]:
# Loop through each folder and count the number of .wav files
for folder in folders:
    if os.path.exists(folder):
        count = len([f for f in os.listdir(folder) if f.endswith('.wav')])
        folder_name = os.path.basename(folder)  # Get the folder name (like Asthma, COPD....)
        print(f"Folder: {folder_name}, Count: {count}")
    else:
        print(f"Folder: {folder} does not exist!")

Folder: Asthma, Count: 319
Folder: Bronchiectasis, Count: 64
Folder: Bronchiolitis, Count: 52
Folder: Bronchitis, Count: 24
Folder: COPD, Count: 3471
Folder: Healthy, Count: 517
Folder: Lung Fibrosis, Count: 41
Folder: Plueral Effusion, Count: 27
Folder: Pneumonia, Count: 187
Folder: URTI, Count: 92


In [ ]:
import librosa
import os
import random
import soundfile as sf

In [ ]:
# def pitch_shift(audio_file, output_file, sr=22050, n_steps=1):
#   y, sr = librosa.load(audio_file,sr=sr)
#   y_shifted = librosa.effects.pitch_shift(y, sr=sr, n_steps=n_steps)
#   sf.write(output_file, y_shifted, sr)  # Use soundfile to write the audio file

# def augment_samples(directory,target_count):
#   files = os.listdir(directory)
#   current_count = len(files)

#   while current_count < target_count:
#     files_to_augment = random.choice(files)
#     base_name, ext = os.path.splitext(files_to_augment)
#     augmented_file = os.path.join(directory, f"{base_name}_aug{current_count}{ext}")

#     pitch_shift(os.path.join(directory, files_to_augment), augmented_file)

#     current_count+=1


In [ ]:
# Function to perform pitch shifting
def pitch_shift(audio_file, output_file, sr=22050, n_steps=1):
    try:
        # Load the audio file
        y, sr = librosa.load(audio_file, sr=sr)

        # Apply pitch shifting
        y_shifted = librosa.effects.pitch_shift(y, sr=sr, n_steps=n_steps)

        # Write the augmented file
        sf.write(output_file, y_shifted, sr)
        print(f"Augmented file saved: {output_file}")
    except Exception as e:
        print(f"Error processing {audio_file}: {e}")

In [ ]:
# Function to augment audio samples in a directory
def augment_samples(directory, target_count):
    files = [f for f in os.listdir(directory) if f.endswith('.wav')]  # Filter only '.wav' files
    current_count = len(files)

    # Augment files until we reach the target count
    while current_count < target_count:
        file_to_augment = random.choice(files)

        # Get base name and extension of the chosen file
        base_name, ext = os.path.splitext(file_to_augment)

        # Skip files that have already been augmented (those with '_aug' in the name)
        if "_aug" in base_name:
            continue

        # Create the new augmented file name
        augmented_file = os.path.join(directory, f"{base_name}_aug{current_count}{ext}")

        # Perform pitch shifting augmentation
        pitch_shift(os.path.join(directory, file_to_augment), augmented_file)

        # Update the count after saving the new file
        current_count += 1

    print(f"Augmentation complete. Total files in {directory}: {current_count}")

### Augmentation FIles to increase to 200
##### Bronchiectasis: From 64 to 200 samples (need 136 more)
##### Bronchiolitis: From 52 to 200 samples (need 148 more)
##### Bronchitis: From 24 to 200 samples (need 176 more)
##### Lung Fibrosis: From 41 to 200 samples (need 159 more)
##### Pleural Effusion: From 27 to 200 samples (need 173 more)
##### Pneumonia: From 187 to 200 samples (need 13 more)
##### URTI: From 92 to 200 samples (need 108 more)

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis/201_1b2_Al_sc_Meditron_segment_0_aug64.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis/201_1b3_Ar_sc_Meditron_segment_0_aug65.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis/201_1b1_Al_sc_Meditron_segment_2_aug66.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis/116_1b2_Tc_sc_Meditron_segment_2_aug67.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis/111_1b2_Tc_sc_Meditron_segment_3_aug68.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiectasis/201_1b2_Ar_sc_Meditron_segment_2_aug69.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound F

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis/149_1b1_Pl_sc_Meditron_segment_3_aug52.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis/216_1b1_Pl_sc_Meditron_segment_2_aug53.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis/216_1b1_Pl_sc_Meditron_segment_2_aug54.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis/206_1b1_Ar_sc_Meditron_segment_3_aug55.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis/161_1b1_Pl_sc_Meditron_segment_2_aug56.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchiolitis/161_1b1_Pl_sc_Meditron_segment_1_aug57.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files D

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis/EP21_BRON,Crep,P R L ,20,M_segment_0_aug24.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis/DP22_BRON,Crep,P R U,24,M_segment_0_aug25.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis/EP21_BRON,Crep,P R L ,20,M_segment_0_aug26.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis/DP28_BRON,Crep,P L U,68,F_segment_1_aug27.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis/DP21_BRON,Crep,P R L ,20,M_segment_0_aug28.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Bronchitis/BP21_BRON,Crep,P R L ,20,M_segment_2_aug29.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Fil

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis/DP23_Lung Fibrosis,Crep,P R L ,50,M_segment_1_aug41.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis/BP23_Lung Fibrosis,Crep,P R L ,50,M_segment_0_aug42.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis/EP26_Lung Fibrosis,Crep,P,90,F_segment_0_aug43.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis/BP23_Lung Fibrosis,Crep,P R L ,50,M_segment_1_aug44.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis/DP105_Lung Fibrosis,Crep,A U R,44,M_segment_4_aug45.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Lung Fibrosis/DP105_Lung Fibrosis,Crep,A U R,44,M_segment_1_aug46.wav
Augmented file 

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion/BP6_Plueral Effusion,I C B,P L R,81,M_segment_4_aug27.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion/DP6_Plueral Effusion,I C B,P L R,81,M_segment_4_aug28.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion/EP6_Plueral Effusion,I C B,P L R,81,M_segment_3_aug29.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion/EP8_Plueral Effusion,I C B,P L L,70,M_segment_1_aug30.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion/EP6_Plueral Effusion,I C B,P L R,81,M_segment_1_aug31.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Plueral Effusion/DP8_Plueral Effusion,I C B,P L L,70,M_

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia/135_2b3_Pr_mc_LittC2SE_segment_0_aug187.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia/EP18_pneumonia,C,P R U,57,M_segment_3_aug188.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia/122_2b3_Ar_mc_LittC2SE_segment_2_aug189.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia/135_2b1_Tc_mc_LittC2SE_segment_2_aug190.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia/122_2b1_Ar_mc_LittC2SE_segment_0_aug191.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumonia/122_2b3_Ar_mc_LittC2SE_segment_2_aug192.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Pneumo

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI', 200)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/210_1b1_Ar_sc_Meditron_segment_0_aug92.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/129_1b1_Ar_sc_Meditron_segment_2_aug93.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/188_1b1_Tc_sc_Meditron_segment_0_aug94.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/150_1b2_Al_sc_Meditron_segment_0_aug95.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/197_1b1_Tc_sc_Meditron_segment_0_aug96.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/150_1b2_Al_sc_Meditron_segment_0_aug97.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/URTI/188_1b1_Pl_sc_Meditron_segment_2_aug98.wav

In [ ]:
augment_samples('/content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy', 1900)

Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/BP96_N,N,P R L ,66,F_segment_1_aug517.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/EP98_N,N,P L L,53,M_segment_0_aug518.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/121_1b1_Tc_sc_Meditron_segment_2_aug519.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/DP72_N,N,P R U,24,F_segment_3_aug520.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/183_1b1_Tc_sc_Meditron_segment_0_aug521.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/DP83_N,N,A R U,18,F_segment_2_aug522.wav
Augmented file saved: /content/drive/MyDrive/Colab Notebooks/Datasets/Lung Sound Files Dataset/Healthy/EP30_N,N,P R M,18,F_segmen

### Check Sound files count after augmentation

In [ ]:
# Loop through each folder and count the number of .wav files
for folder in folders:
    if os.path.exists(folder):
        count = len([f for f in os.listdir(folder) if f.endswith('.wav')])
        folder_name = os.path.basename(folder)  # Get the folder name (like Asthma, COPD, etc.)
        print(f"Folder: {folder_name}, Count: {count}")
    else:
        print(f"Folder: {folder} does not exist!")

Folder: Asthma, Count: 319
Folder: Bronchiectasis, Count: 200
Folder: Bronchiolitis, Count: 200
Folder: Bronchitis, Count: 200
Folder: COPD, Count: 3471
Folder: Healthy, Count: 1900
Folder: Lung Fibrosis, Count: 200
Folder: Plueral Effusion, Count: 200
Folder: Pneumonia, Count: 200
Folder: URTI, Count: 200
